src/ucp/core/ucp_worker.c

```diff
@@ -533,65 +454,56 @@ ucs_status_t ucp_worker_set_ep_failed(ucp_worker_h worker, ucp_ep_h ucp_ep,
         goto out_ok;
     }
 
-    /* Release EP ID here to prevent protocols from sending reply */
-    ucp_ep_release_id(ucp_ep);
-    ucp_ep_update_flags(ucp_ep, UCP_EP_FLAG_FAILED, 0);
-
     if (ucp_ep_config(ucp_ep)->key.err_mode == UCP_ERR_HANDLING_MODE_NONE) {
         /* NOTE: if user has not requested error handling on the endpoint,
          *       the failure is considered unhandled */
+        ucp_ep_update_flags(ucp_ep, UCP_EP_FLAG_FAILED, 0);
+        /* Release EP ID here to prevent protocols from sending reply */
+        ucp_ep_release_id(ucp_ep);
```

Comment:
why moved these lines inside the if?

Reply:
no need to do it if error mode is not equal `NONE`.
setting FAILED flag and releasing ID is done in `ucp_ep_discard_lanes()`

Reply:
would be better to do it only in one place

Reply:
but we should do it in `set_lanes_failed` (which is a part of `discard_lanes`) too

Reply:
maybe we can do set lanes failed also when err handling not enabled?
if we have less code difference between err handling on and off it will be easier to maintain

Reply:
discussed: change ucp_ep_discard_lanes to not pass FLAG CANCEL (pass FLAG LOCAL instead)  if no err handling is enabled

Reply:
done